pkg/worker/dead_jobs_handler.go

```diff
@@ -0,0 +1,132 @@
+package worker
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+	"html/template"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+//go:embed dead_jobs_page.html
+var deadJobsPageTplSrc string
+
+//go:generate mockery --name=DeadJobManager -r --case underscore --with-expecter --structname DeadJobManager --filename dead_job_manager_mock.go --output=./mocks
+
+type DeadJobManager interface {
+	DeadJobs(ctx context.Context, size, offset int) ([]Job, error)
+	Resurrect(ctx context.Context, jobIDs []string) error
+	ClearDeadJobs(ctx context.Context, jobIDs []string) error
+}
+
+const (
+	listDeadJobsPath  = "/dead-jobs"
+	resurrectJobsPath = "/resurrect-jobs"
+	clearJobsPath     = "/clear-jobs"
+)
+
+// DeadJobManagementHandler returns a http handler with endpoints for dead job
+// management:
+//   - /dead-jobs: JSON/HTML response with content negotiation. Response is
+//     paginated.
+//   - /resurrect-jobs: Move specified dead jobs to jobs_queue table.
+//   - /clear-jobs: Remove specified dead jobs from dead_jobs table.
+func DeadJobManagementHandler(mgr DeadJobManager) http.Handler {
+	mux := http.NewServeMux()
+	mux.HandleFunc(listDeadJobsPath, deadJobsHandler(mgr))
+	mux.HandleFunc(resurrectJobsPath, clearOrResurrectHandler(mgr, "resurrect"))
+	mux.HandleFunc(clearJobsPath, clearOrResurrectHandler(mgr, "clear"))
+	return mux
+}
+
+func deadJobsHandler(mgr DeadJobManager) http.HandlerFunc {
+	deadJobsPageTpl := template.Must(template.New("").Parse(deadJobsPageTplSrc))
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		qry := r.URL.Query()
+		size, err := strconv.Atoi(qry.Get("size"))
+		if err != nil || size <= 0 {
+			size = 20
+		}
+
+		offset, _ := strconv.Atoi(qry.Get("offset"))
+		if offset <= 0 {
+			offset = 0
+		}
+
+		jobs, err := mgr.DeadJobs(r.Context(), size, offset)
+		if err != nil {
+			writeJSONResponse(w, http.StatusInternalServerError, err)
+			return
+		}
+
+		if strings.Contains(r.Header.Get("Accept"), "application/json") {
+			writeJSONResponse(w, http.StatusOK, jobs)
+			return
+		}
+
+		tplData := map[string]any{
+			"jobs":      jobs,
+			"page_size": size,
+			"next_page": offset + size,
+			"prev_page": offset - size,
+			"err_msg":   strings.TrimSpace(qry.Get("error")),
+		}
+		w.Header().Set("Content-Type", "text/html")
+		w.WriteHeader(http.StatusOK)
+		if err := deadJobsPageTpl.Execute(w, tplData); err != nil {
+			sendTo := listDeadJobsPath + "?error=" + url.QueryEscape(err.Error())
+			http.Redirect(w, r, sendTo, http.StatusSeeOther)
+			return
+		}
+	}
+}
+
+func clearOrResurrectHandler(mgr DeadJobManager, mode string) http.HandlerFunc {
```

Comment:
can we segregate this to clearDeadJobsHandler and resurrectDeadJobsHandler?

Reply:
We had initially kept the handler separate but the handlers were virtually identical so we decided to write it as `clearOrResurrectHandler`. Do we really need to split it?

Reply:
yes I think injecting params in handlers is better be avoided, identical handler is fine, meaning we can also segregate the test cases too to avoid mixing things up

Reply:
Could you elaborate on why you think it would be better to avoid the `mode` param?

The test cases are already segregated - they aren't dependent on implementation, just tests the API that is the exposed endpoints.

Reply:
we should consider the single responsibility for each handler. Injecting mode in the handler and separating the logic inside will make future maintenance complicated (e.g. adding more handler with identical validation but different use case). Looking at this handler, default mode is to clear the dead jobs, any typo done on passing the mode will result in clearing the dead jobs (even tho it's highly unlikely to do typo). This cannot be the case if we explicitly segregate the handler. If we still want to use this approach, I'd suggest to also validate valid modes inside the handler.

If we're concerned on the redundancy, we can write common function used by both handlers instead of merging both handler into 1 func.

Reply:
I have refactored the handler, please check it.

Reply:
yep this is also fine approach